lib/msal-common/src/response/ResponseHandler.ts

```diff
@@ -93,39 +137,34 @@ export class ResponseHandler {
         if (idTokenObj.claims.nonce !== nonce) {
             throw ClientAuthError.createNonceMismatchError();
         }
-    
-        // TODO: This will be used when saving tokens
-        // const authorityKey: string = this.cacheManager.generateAuthorityKey(state);
-        // const cachedAuthority: string = this.cacheStorage.getItem(authorityKey);
-    
-        // TODO: Save id token here
-        
+
+        this.cacheStorage.setItem(PersistentCacheKeys.ID_TOKEN, idTokenObj.rawIdToken);
+
+        const authorityKey: string = this.cacheManager.generateAuthorityKey(state);
+        const cachedAuthority: string = this.cacheStorage.getItem(authorityKey);
+
         // Retrieve client info
         const clientInfo = buildClientInfo(this.cacheStorage.getItem(PersistentCacheKeys.CLIENT_INFO), this.cryptoObj);
-    
+
         // Create account object for request
         tokenResponse.account = Account.createAccount(idTokenObj, clientInfo, this.cryptoObj);
-    
+
         // Set token type
         tokenResponse.tokenType = serverTokenResponse.token_type;
-    
+
         // Save the access token if it exists
-        if (serverTokenResponse.access_token) {
-            const accountKey = this.cacheManager.generateAcquireTokenAccountKey(tokenResponse.account.homeAccountIdentifier);
-            
-            const cachedAccount = JSON.parse(this.cacheStorage.getItem(accountKey)) as Account;
-
-            if (!cachedAccount || Account.compareAccounts(cachedAccount, tokenResponse.account)) {
-                tokenResponse.accessToken = serverTokenResponse.access_token;
-                tokenResponse.refreshToken = serverTokenResponse.refresh_token;
-                // TODO: Save the access token
-            } else {
-                throw ClientAuthError.createAccountMismatchError();
-            }
+        const accountKey = this.cacheManager.generateAcquireTokenAccountKey(tokenResponse.account.homeAccountIdentifier);
+        const cachedAccount = this.getCachedAccount(accountKey);
+        if (!cachedAccount || Account.compareAccounts(cachedAccount, tokenResponse.account)) {
+            this.saveToken(tokenResponse, cachedAuthority, serverTokenResponse, clientInfo);
+        } else if (cachedAccount) {
+            throw ClientAuthError.createAccountMismatchError();
         }
 
         // Return user set state in the response
         tokenResponse.userRequestState = ProtocolUtils.getUserRequestState(state);
+        
+        this.cacheManager.resetTempCacheItems(state);
```

Comment:
Same logic as Jason's comment above, constructing each param and building tokenResponse at the end of the function makes it more readable.

Reply:
Updated the flow, will update more in a future PR since this is used for renewToken flows as well.